<!DOCTYPE html>
<html lang="en">
<head>
  <link href='//fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700,400italic' rel='stylesheet' type='text/css'>

    <link rel="stylesheet" type="text/css" href="http://xion.io/$theme/stylesheet/style.min.css">

  <link rel="stylesheet" type="text/css" href="http://xion.io/$theme/stylesheet/pygments.min.css">
  <link rel="stylesheet" type="text/css" href="http://xion.io/$theme/stylesheet/font-awesome.min.css">

    <link href="http://xion.io/style.css" rel="stylesheet">




  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta name="robots" content="" />

<meta name="author" content="Karol Kuczmarski" />
<meta name="description" content="Here’s a little trick that may be useful in dealing with asynchronous Streams in Rust. When you consume a Stream using the for_each method, its default behavior is to finish early should an error be produced by the stream: use futures::prelude::*; use futures::stream; use tokio_core::reactor::Core; let s = stream::iter_result(vec![Ok(1), Ok(2), Err(false), Ok(3)]); let fut = s.for_each(|n| { println!(&#34;{}&#34;, n); Ok(()) }); In more precise terms, it means that the Future returned by for_each will resolve with the first error from the underlying stream: // Prints 1, 2, and then panics with &#34;false&#34;. Core::new().unwrap().run(fut).unwrap(); For most purposes, this is perfectly alright; errors are generally meant to propagate, after all. Certain kinds of errors, however, are better off silenced. Perhaps they are expected to pop up during normal program operation, or maybe their occurrence should merely affect program execution in a particular way, and not halt it outright. In a simple case like above, you can of course check what for_each itself has returned, but that doesn’t scale to building larger Stream pipelines. I encountered a situation like this myself when using the hubcaps library. The code I was writing was meant to search for GitHub issues within a specific repository. In GitHub API, this is accomplished by sending a search query like repo:$OWNER/$NAME, which may result in a rather obscure HTTP error (422 Unprocessable Entity) if the given repository doesn’t actually exist. But I didn’t care about this error; should it occur, I’d simply return an empty stream, because doing so was more convenient for the larger bit of logic that was consuming it. Unfortunately, the Stream trait offers no interface that’d target this use case. There are only a few methods that even allow to look at errors mid-stream, and even fewer that can end it prematurely. On the flip side, at least we don’t have to consider too many combinations when looking for the solution ;) Indeed, it seems there are only two Stream methods that are worthy of our attention: Stream::then, because it allows for a closure to receive all stream values (items and errors) Stream::take_while, because it accepts a closure that can end the stream early (but only based on items, not errors) Combining them both, we arrive at the following recipe: Inside a .then closure, look for Errors that you consider non-fatal and replace them with a special item value. The natural choice for such a value is None. As a side effect, this forces us to convert the regular (“successful”) items into Some(item), effectively transforming a Stream&lt;Item=T&gt; into Stream&lt;Item=Option&lt;T&gt;&gt;. Looks for the special value (i.e. None) in the .take_while closure and terminate the stream when it’s been found. Finally, convert the wrapped items back into their original form using .map, thus giving us back a Stream of T‘s. Applying this technique to our initial example, we get something that looks like this: let s = stream::iter_result(vec![Ok(1), Ok(2), Err(false), Ok(3)]) .then(|r| match r { Ok(r) =&gt; Ok(Some(r)), // no-op passthrough of items Err(false) =&gt; Ok(None) // non-fatal error, terminate the stream Err(e) =&gt; Err(e), // no-op passthrough of other errors }) .take_while(|x| future::ok(x.is_some())) .map(Option::unwrap); If we now try to consume this stream like before: Core::new().run( s.for_each(|n| { println!(&#34;{}&#34;, n); Ok(()) }) ).unwrap(); it will still end after the first two items, but without producing any errors afterwards. For a more reusable version of the trick, you can check this gist; it adds a Stream::take_while_err method through an extension trait. This isn’t a perfect solution, however, because it requires Boxing even on nightly Rust1. We can fix that by introducing a dedicated TakeWhileErr stream type, similarly to what native Stream methods do. I leave that as an exercise for the reader ;-) This is due to a limitation in the impl Trait feature which prevents it from being used as a return type of trait methods. ↩" />
<meta name="keywords" content="Rust, streams, Tokio, async">
<meta property="og:site_name" content="Karol Kuczmarski's Blog"/>
<meta property="og:title" content="Terminating a Stream in Rust"/>
<meta property="og:description" content="Here’s a little trick that may be useful in dealing with asynchronous Streams in Rust. When you consume a Stream using the for_each method, its default behavior is to finish early should an error be produced by the stream: use futures::prelude::*; use futures::stream; use tokio_core::reactor::Core; let s = stream::iter_result(vec![Ok(1), Ok(2), Err(false), Ok(3)]); let fut = s.for_each(|n| { println!(&#34;{}&#34;, n); Ok(()) }); In more precise terms, it means that the Future returned by for_each will resolve with the first error from the underlying stream: // Prints 1, 2, and then panics with &#34;false&#34;. Core::new().unwrap().run(fut).unwrap(); For most purposes, this is perfectly alright; errors are generally meant to propagate, after all. Certain kinds of errors, however, are better off silenced. Perhaps they are expected to pop up during normal program operation, or maybe their occurrence should merely affect program execution in a particular way, and not halt it outright. In a simple case like above, you can of course check what for_each itself has returned, but that doesn’t scale to building larger Stream pipelines. I encountered a situation like this myself when using the hubcaps library. The code I was writing was meant to search for GitHub issues within a specific repository. In GitHub API, this is accomplished by sending a search query like repo:$OWNER/$NAME, which may result in a rather obscure HTTP error (422 Unprocessable Entity) if the given repository doesn’t actually exist. But I didn’t care about this error; should it occur, I’d simply return an empty stream, because doing so was more convenient for the larger bit of logic that was consuming it. Unfortunately, the Stream trait offers no interface that’d target this use case. There are only a few methods that even allow to look at errors mid-stream, and even fewer that can end it prematurely. On the flip side, at least we don’t have to consider too many combinations when looking for the solution ;) Indeed, it seems there are only two Stream methods that are worthy of our attention: Stream::then, because it allows for a closure to receive all stream values (items and errors) Stream::take_while, because it accepts a closure that can end the stream early (but only based on items, not errors) Combining them both, we arrive at the following recipe: Inside a .then closure, look for Errors that you consider non-fatal and replace them with a special item value. The natural choice for such a value is None. As a side effect, this forces us to convert the regular (“successful”) items into Some(item), effectively transforming a Stream&lt;Item=T&gt; into Stream&lt;Item=Option&lt;T&gt;&gt;. Looks for the special value (i.e. None) in the .take_while closure and terminate the stream when it’s been found. Finally, convert the wrapped items back into their original form using .map, thus giving us back a Stream of T‘s. Applying this technique to our initial example, we get something that looks like this: let s = stream::iter_result(vec![Ok(1), Ok(2), Err(false), Ok(3)]) .then(|r| match r { Ok(r) =&gt; Ok(Some(r)), // no-op passthrough of items Err(false) =&gt; Ok(None) // non-fatal error, terminate the stream Err(e) =&gt; Err(e), // no-op passthrough of other errors }) .take_while(|x| future::ok(x.is_some())) .map(Option::unwrap); If we now try to consume this stream like before: Core::new().run( s.for_each(|n| { println!(&#34;{}&#34;, n); Ok(()) }) ).unwrap(); it will still end after the first two items, but without producing any errors afterwards. For a more reusable version of the trick, you can check this gist; it adds a Stream::take_while_err method through an extension trait. This isn’t a perfect solution, however, because it requires Boxing even on nightly Rust1. We can fix that by introducing a dedicated TakeWhileErr stream type, similarly to what native Stream methods do. I leave that as an exercise for the reader ;-) This is due to a limitation in the impl Trait feature which prevents it from being used as a return type of trait methods. ↩"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:url" content="http://xion.io/post/code/rust-stream-terminate.html"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2017-12-16 21:17:00+01:00"/>
<meta property="article:modified_time" content=""/>
<meta property="article:author" content="http://xion.io/">
<meta property="article:section" content="Code"/>
<meta property="article:tag" content="Rust"/>
<meta property="article:tag" content="streams"/>
<meta property="article:tag" content="Tokio"/>
<meta property="article:tag" content="async"/>
<meta property="og:image" content="http://xion.io/logo.jpeg">
  <title>Karol Kuczmarski's Blog &ndash; Terminating a Stream in Rust</title>
</head>
<body>
  <aside>
    <div>
      <a href="http://xion.io">
        <img src="http://xion.io/logo.jpeg" alt="Karol Kuczmarski" title="Karol Kuczmarski">
      </a>
      <h1><a href="http://xion.io">Karol Kuczmarski</a></h1>
      <p>fn(Tea) -> Code</p>
      <nav>
        <ul class="list">
          <li><a href="http://xion.io/page/about.html#about">About</a></li>
          <li><a href="http://xion.io/page/projects.html#projects">Projects</a></li>
        </ul>
      </nav>
      <ul class="social">
        <li><a class="sc-github" href="http://github.com/Xion" target="_blank"><i class="fa fa-github"></i></a></li>
        <li><a class="sc-stack-overflow" href="http://stackoverflow.com/users/434799/xion" target="_blank"><i class="fa fa-stack-overflow"></i></a></li>
        <li><a class="sc-twitter" href="http://twitter.com/Xion__" target="_blank"><i class="fa fa-twitter"></i></a></li>
        <li><a class="sc-google" href="https://plus.google.com/+KarolKuczmarski" target="_blank"><i class="fa fa-google"></i></a></li>
        <li><a class="sc-rss" href="/feeds/atom.xml" target="_blank"><i class="fa fa-rss"></i></a></li>
      </ul>
    </div>
  </aside>
  <main>
    <nav>
      <a href="http://xion.io">Home</a>
      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>
      <a href="http://xion.org.pl/">Old blog</a>
    </nav>

<article>
  <header>
    <h1 id="rust-stream-terminate">Terminating a Stream in&nbsp;Rust</h1>
    <p>Posted on Sat 16 December 2017 in <a href="http://xion.io/category/code.html">Code</a></p>
  </header>
  <div>
    <p>Here&#8217;s a little trick that may be useful in dealing with
<a href="https://docs.rs/futures/0.1.17/futures/stream/trait.Stream.html">asynchronous <code>Stream</code>s</a> in&nbsp;Rust.</p>
<p>When you consume a <code>Stream</code> using
<a href="https://docs.rs/futures/0.1.17/futures/stream/trait.Stream.html#method.for_each">the <code>for_each</code> method</a>,
its default behavior is to finish early should an error be produced by the&nbsp;stream:</p>
<div class="highlight"><pre><span class="kn">use</span><span class="w"> </span><span class="n">futures</span><span class="o">::</span><span class="n">prelude</span><span class="o">::*</span><span class="p">;</span><span class="w"></span>
<span class="kn">use</span><span class="w"> </span><span class="n">futures</span><span class="o">::</span><span class="n">stream</span><span class="p">;</span><span class="w"></span>
<span class="kn">use</span><span class="w"> </span><span class="n">tokio_core</span><span class="o">::</span><span class="n">reactor</span><span class="o">::</span><span class="n">Core</span><span class="p">;</span><span class="w"></span>

<span class="kd">let</span><span class="w"> </span><span class="n">s</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">stream</span><span class="o">::</span><span class="n">iter_result</span><span class="p">(</span><span class="n">vec</span><span class="o">!</span><span class="p">[</span><span class="nb">Ok</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span><span class="w"> </span><span class="nb">Ok</span><span class="p">(</span><span class="mi">2</span><span class="p">),</span><span class="w"> </span><span class="nb">Err</span><span class="p">(</span><span class="kc">false</span><span class="p">),</span><span class="w"> </span><span class="nb">Ok</span><span class="p">(</span><span class="mi">3</span><span class="p">)]);</span><span class="w"></span>
<span class="kd">let</span><span class="w"> </span><span class="n">fut</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">s</span><span class="p">.</span><span class="n">for_each</span><span class="p">(</span><span class="o">|</span><span class="n">n</span><span class="o">|</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">    </span><span class="nb">println</span><span class="o">!</span><span class="p">(</span><span class="s">&quot;{}&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="p">);</span><span class="w"></span>
<span class="w">    </span><span class="nb">Ok</span><span class="p">(())</span><span class="w"></span>
<span class="p">});</span><span class="w"></span>
</pre></div>


<p>In more precise terms, it means that
<a href="https://docs.rs/futures/0.1.17/futures/future/trait.Future.html">the <code>Future</code></a>
returned by <code>for_each</code> will resolve with the first error
from the underlying&nbsp;stream:</p>
<div class="highlight"><pre><span class="c1">// Prints 1, 2, and then panics with &quot;false&quot;.</span>
<span class="n">Core</span><span class="o">::</span><span class="n">new</span><span class="p">().</span><span class="n">unwrap</span><span class="p">().</span><span class="n">run</span><span class="p">(</span><span class="n">fut</span><span class="p">).</span><span class="n">unwrap</span><span class="p">();</span><span class="w"></span>
</pre></div>


<p>For most purposes, this is perfectly alright;
errors are generally meant to propagate, after&nbsp;all.</p>
<p>Certain <em>kinds</em> of errors, however, are better off silenced.
Perhaps they are expected to pop up during normal program operation,
or maybe their occurrence should merely <em>affect</em> program execution in a particular way,
and not halt it outright.
In a simple case like above, you can of course check what <code>for_each</code> itself has returned,
but that doesn&#8217;t scale to building larger <code>Stream</code> pipelines.</p>
<p>I encountered a situation like this myself when using
<a href="https://docs.rs/crate/hubcaps/0.4.2">the <em>hubcaps</em> library</a>.
The code I was writing was meant to
<a href="https://docs.rs/hubcaps/0.4.2/hubcaps/search/struct.SearchIssues.html#method.iter">search for GitHub issues</a>
within a specific repository.
In GitHub <span class="caps">API</span>, this is accomplished by sending a search query like <code>repo:$OWNER/$NAME</code>,
which may result in a rather obscure <span class="caps">HTTP</span> error (<a href="https://httpstatuses.com/422">422 Unprocessable Entity</a>)
if the given repository doesn&#8217;t actually exist.
But I didn&#8217;t care about this error; should it occur, I&#8217;d simply return an empty stream,
because doing so was more convenient for the larger bit of logic that was consuming&nbsp;it.</p>
<p>Unfortunately, the <code>Stream</code> trait offers no interface that&#8217;d target this use case.
There are only a few methods that even allow to look at errors mid-stream,
and even fewer that can end it prematurely.
On the flip side, at least we don&#8217;t have to consider <em>too</em> many combinations
when looking for the solution&nbsp;;)</p>
<p>Indeed, it seems there are only two <code>Stream</code> methods that are worthy of our&nbsp;attention:</p>
<ul>
<li><a href="https://docs.rs/futures/0.1.17/futures/stream/trait.Stream.html#method.then"><code>Stream::then</code></a>,
  because it allows for a closure to receive all stream values (items <em>and</em>&nbsp;errors)</li>
<li><a href="https://docs.rs/futures/0.1.17/futures/stream/trait.Stream.html#method.take_while"><code>Stream::take_while</code></a>,
  because it accepts a closure that can end the stream early (but only based on items, not&nbsp;errors)</li>
</ul>
<p>Combining them both, we arrive at the following&nbsp;recipe:</p>
<ul>
<li>
<p>Inside a <code>.then</code> closure, look for <code>Err</code>ors that you consider non-fatal
  and replace them with a special item value.
  The natural choice for such a value is <code>None</code>.
  As a side effect, this forces us to convert the regular (&#8220;successful&#8221;) <code>item</code>s into <code>Some(item)</code>,
  effectively transforming a <code>Stream&lt;Item=T&gt;</code> into <code>Stream&lt;Item=Option&lt;T&gt;&gt;</code>.</p>
</li>
<li>
<p>Looks for the special value (i.e. <code>None</code>) in the <code>.take_while</code> closure
  and terminate the stream when it&#8217;s been&nbsp;found.</p>
</li>
<li>
<p>Finally, convert the wrapped items back into their original form using <code>.map</code>,
  thus giving us back a <code>Stream</code> of <code>T</code><span class="quo">&#8216;</span>s.</p>
</li>
</ul>
<p>Applying this technique to our initial example,
we get something that looks like&nbsp;this:</p>
<div class="highlight"><pre><span class="kd">let</span><span class="w"> </span><span class="n">s</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">stream</span><span class="o">::</span><span class="n">iter_result</span><span class="p">(</span><span class="n">vec</span><span class="o">!</span><span class="p">[</span><span class="nb">Ok</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span><span class="w"> </span><span class="nb">Ok</span><span class="p">(</span><span class="mi">2</span><span class="p">),</span><span class="w"> </span><span class="nb">Err</span><span class="p">(</span><span class="kc">false</span><span class="p">),</span><span class="w"> </span><span class="nb">Ok</span><span class="p">(</span><span class="mi">3</span><span class="p">)])</span><span class="w"></span>
<span class="w">    </span><span class="p">.</span><span class="n">then</span><span class="p">(</span><span class="o">|</span><span class="n">r</span><span class="o">|</span><span class="w"> </span><span class="k">match</span><span class="w"> </span><span class="n">r</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">        </span><span class="nb">Ok</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w"> </span><span class="o">=&gt;</span><span class="w"> </span><span class="nb">Ok</span><span class="p">(</span><span class="nb">Some</span><span class="p">(</span><span class="n">r</span><span class="p">)),</span><span class="w">  </span><span class="c1">// no-op passthrough of items</span>
<span class="w">        </span><span class="nb">Err</span><span class="p">(</span><span class="kc">false</span><span class="p">)</span><span class="w"> </span><span class="o">=&gt;</span><span class="w"> </span><span class="nb">Ok</span><span class="p">(</span><span class="nb">None</span><span class="p">)</span><span class="w"> </span><span class="c1">// non-fatal error, terminate the stream</span>
<span class="w">        </span><span class="nb">Err</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="w"> </span><span class="o">=&gt;</span><span class="w"> </span><span class="nb">Err</span><span class="p">(</span><span class="n">e</span><span class="p">),</span><span class="w">      </span><span class="c1">// no-op passthrough of other errors</span>
<span class="w">    </span><span class="p">})</span><span class="w"></span>
<span class="w">    </span><span class="p">.</span><span class="n">take_while</span><span class="p">(</span><span class="o">|</span><span class="n">x</span><span class="o">|</span><span class="w"> </span><span class="n">future</span><span class="o">::</span><span class="n">ok</span><span class="p">(</span><span class="n">x</span><span class="p">.</span><span class="n">is_some</span><span class="p">()))</span><span class="w"></span>
<span class="w">    </span><span class="p">.</span><span class="n">map</span><span class="p">(</span><span class="nb">Option</span><span class="o">::</span><span class="n">unwrap</span><span class="p">);</span><span class="w"></span>
</pre></div>


<p>If we now try to consume this stream like&nbsp;before:</p>
<div class="highlight"><pre><span class="n">Core</span><span class="o">::</span><span class="n">new</span><span class="p">().</span><span class="n">run</span><span class="p">(</span><span class="w"></span>
<span class="w">    </span><span class="n">s</span><span class="p">.</span><span class="n">for_each</span><span class="p">(</span><span class="o">|</span><span class="n">n</span><span class="o">|</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="nb">println</span><span class="o">!</span><span class="p">(</span><span class="s">&quot;{}&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="p">);</span><span class="w"> </span><span class="nb">Ok</span><span class="p">(())</span><span class="w"> </span><span class="p">})</span><span class="w"></span>
<span class="p">).</span><span class="n">unwrap</span><span class="p">();</span><span class="w"></span>
</pre></div>


<p>it will still end after the first two items,
but without producing any errors&nbsp;afterwards.</p>
<hr />
<p>For a more reusable version of the trick, you can check
<a href="https://play.rust-lang.org/?gist=1bf2199460258be6674c9c7a7a157f50&amp;version=stable">this gist</a>;
it adds a <code>Stream::take_while_err</code> method through an <a href="http://xion.io/post/code/rust-extension-traits.html">extension trait</a>.</p>
<p>This isn&#8217;t a perfect solution, however, because it requires <code>Box</code>ing even on nightly Rust<sup id="fnref:1"><a class="footnote-ref" href="#fn:1" rel="footnote">1</a></sup>.
We can fix that by introducing a dedicated <code>TakeWhileErr</code> stream type,
similarly to what native <code>Stream</code> methods do.
I leave that as an exercise for the reader&nbsp;;-)</p>
<div class="footnote">
<hr />
<ol>
<li id="fn:1">
<p>This is due to a limitation in <a href="https://github.com/rust-lang/rust/issues/34511">the <code>impl Trait</code> feature</a>
which prevents it from being used as a return type of trait methods.&#160;<a class="footnote-backref" href="#fnref:1" rev="footnote" title="Jump back to footnote 1 in the text">&#8617;</a></p>
</li>
</ol>
</div>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="http://xion.io/tag/rust.html">Rust</a>
      <a href="http://xion.io/tag/streams.html">streams</a>
      <a href="http://xion.io/tag/tokio.html">Tokio</a>
      <a href="http://xion.io/tag/async.html">async</a>
    </p>
  </div>
<div id="disqus_thread"></div>
<script type="text/javascript">
    var disqus_shortname = 'xionblog';
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript>
</article>

    <footer>
<p>
  &copy; Karol Kuczmarski 2018 - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>
</p>
<p>Built using <a href="http://getpelican.com" target="_blank">Pelican</a> - <a href="https://github.com/alexandrevicenzi/flex" target="_blank">Flex</a> theme by <a href="http://alexandrevicenzi.com" target="_blank">Alexandre Vicenzi</a></p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
         src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>

<!-- Google Analytics -->
<script type="text/javascript">
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-27379564-1', 'auto');
  ga('send', 'pageview');
</script>
<!-- End Google Analytics -->



<script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "BlogPosting",
  "name": "Terminating a Stream in Rust",
  "headline": "Terminating a Stream in Rust",
  "datePublished": "2017-12-16 21:17:00+01:00",
  "dateModified": "",
  "author": {
    "@type": "Person",
    "name": "Karol Kuczmarski",
    "url": "http://xion.io/"
  },
  "image": "http://xion.io/logo.jpeg",
  "url": "http://xion.io/post/code/rust-stream-terminate.html",
  "description": "Here’s a little trick that may be useful in dealing with asynchronous Streams in Rust. When you consume a Stream using the for_each method, its default behavior is to finish early should an error be produced by the stream: use futures::prelude::*; use futures::stream; use tokio_core::reactor::Core; let s = stream::iter_result(vec![Ok(1), Ok(2), Err(false), Ok(3)]); let fut = s.for_each(|n| { println!("{}", n); Ok(()) }); In more precise terms, it means that the Future returned by for_each will resolve with the first error from the underlying stream: // Prints 1, 2, and then panics with "false". Core::new().unwrap().run(fut).unwrap(); For most purposes, this is perfectly alright; errors are generally meant to propagate, after all. Certain kinds of errors, however, are better off silenced. Perhaps they are expected to pop up during normal program operation, or maybe their occurrence should merely affect program execution in a particular way, and not halt it outright. In a simple case like above, you can of course check what for_each itself has returned, but that doesn’t scale to building larger Stream pipelines. I encountered a situation like this myself when using the hubcaps library. The code I was writing was meant to search for GitHub issues within a specific repository. In GitHub API, this is accomplished by sending a search query like repo:$OWNER/$NAME, which may result in a rather obscure HTTP error (422 Unprocessable Entity) if the given repository doesn’t actually exist. But I didn’t care about this error; should it occur, I’d simply return an empty stream, because doing so was more convenient for the larger bit of logic that was consuming it. Unfortunately, the Stream trait offers no interface that’d target this use case. There are only a few methods that even allow to look at errors mid-stream, and even fewer that can end it prematurely. On the flip side, at least we don’t have to consider too many combinations when looking for the solution ;) Indeed, it seems there are only two Stream methods that are worthy of our attention: Stream::then, because it allows for a closure to receive all stream values (items and errors) Stream::take_while, because it accepts a closure that can end the stream early (but only based on items, not errors) Combining them both, we arrive at the following recipe: Inside a .then closure, look for Errors that you consider non-fatal and replace them with a special item value. The natural choice for such a value is None. As a side effect, this forces us to convert the regular (“successful”) items into Some(item), effectively transforming a Stream<Item=T> into Stream<Item=Option<T>>. Looks for the special value (i.e. None) in the .take_while closure and terminate the stream when it’s been found. Finally, convert the wrapped items back into their original form using .map, thus giving us back a Stream of T‘s. Applying this technique to our initial example, we get something that looks like this: let s = stream::iter_result(vec![Ok(1), Ok(2), Err(false), Ok(3)]) .then(|r| match r { Ok(r) => Ok(Some(r)), // no-op passthrough of items Err(false) => Ok(None) // non-fatal error, terminate the stream Err(e) => Err(e), // no-op passthrough of other errors }) .take_while(|x| future::ok(x.is_some())) .map(Option::unwrap); If we now try to consume this stream like before: Core::new().run( s.for_each(|n| { println!("{}", n); Ok(()) }) ).unwrap(); it will still end after the first two items, but without producing any errors afterwards. For a more reusable version of the trick, you can check this gist; it adds a Stream::take_while_err method through an extension trait. This isn’t a perfect solution, however, because it requires Boxing even on nightly Rust1. We can fix that by introducing a dedicated TakeWhileErr stream type, similarly to what native Stream methods do. I leave that as an exercise for the reader ;-) This is due to a limitation in the impl Trait feature which prevents it from being used as a return type of trait methods. ↩"
}
</script></body>
</html>